src/adapter/src/optimize/dataflow_expiration.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+//! Helper function for dataflow expiration checks.
+//!
+//! A [`TimeDependence`] describes how a dataflow follows wall-clock time, and this module offers
+//! functions to compute the time dependence based on the current state of the catalog.
+//!
+//! For a set of global IDs, the helper determines how a dataflow that depends on the set of IDs
+//! follows wall-clock time. Within the catalog, we remember how all installed objects follow
+//! wall-clock time, which turns this step into gathering and combining different time dependencies:
+//! * A meet of anything with wall-clock time results in wall-clock time.
+//! * A meet of anything but wall-clock time and a refresh schedule results in a refresh schedule
+//!   that depends on the deduplicated collection of dependencies.
+//! * Otherwise, a dataflow is indeterminate, which expresses that we either don't know how it
+//!   follows wall-clock time, or is a constant collection.
+//!
+//! The time dependence needs to be computed on the actual dependencies, and not on catalog
+//! uses. An optimized dataflow depends on concrete indexes, and has unnecessary dependencies
+//! pruned. Additionally, transitive dependencies can depend on indexes that do not exist anymore,
+//! which makes combining run-time information with catalog-based information inconclusive.
```

Comment:
(Note that in theory, it could happen that an MIR plan involves a GlobalId that actually happens to be "dead code" in the MIR plan. In this case, `DataflowDescription::import_ids` would still list this dependency, but the dataflow's actual frontier wouldn't reflect this dependency. In practice, this shouldn't happen currently, because one of the last things that we do with MIR plans is call `NormalizeLets`, which removes CTEs that are not referenced.)